Stateless BOLT 12 `Offer` verification

jkczyz opened the pull request:
An `Offer` may contain metadata which can be used to authenticate that the `Offer` was created from a known key. This allows for stateless verification of an `Offer` when an `InvoiceRequest` is received. The `InvoiceRequest` contains all the TLV records from the `Offer`.

Based on #1972.

@TheBlueMatt  @valentinewallace Looking for early feedback on the HEAD commit. This is simpler than the approach used in `inbound_payments` since the `InvoiceRequest` contains all the `Offer`'s fields. However, I'm not sure if this is sufficient.

The signing pubkey should probably be ephemeral and stateless as well, as it is needed to sign an `Invoice` created from the `InvoiceRequest`. So it seems the corresponding private key will need to be generated using an `EntropySource` and encrypted in the metadata. Does that sound right or do you have an alternative in mind?

jkczyz commented:
Need to check my understanding here...

> Discussed very briefly offline - for metadata verification we, I think, need to be able to sign with both our node and and an auto-derived one.

How should we represent this in the `OfferBuilder` interface? Currently, `OfferBuilder::new` requires a `PublicKey` since `OfferContents::signing_pubkey` is not an `Option` and would need to be initialized to something.

We could make `OfferContents::signing_pubkey` an `Option` and live with unwrapping internally. Then we could have `OfferBuilder::new` take an enum like:

```rust
enum SigningPubkey {
    Explicit(PublicKey),
    Derived,
}
```

Which at very least could be used in the builder to determine if `signing_pubkey` should be set to `None` and derived when calling `build`. Storing the enum directly in `OfferContents` instead would have implications for `OfferContents::signing_pubkey` method, `OfferContents::as_tvl_stream`, writing, and parsing, so I'd want to avoid that at least in favor of `Option<PublicKey>`.

> That means, I think, that we should do the verification of the metadata without the signing key at all, but then also derive the signing key from that metadata if we're not using our node_id.

If a user gives an explicit pubkey and their own metadata, I assume we can't verify then? I guess this is expected and would mean they couldn't use our onion message handler to respond to the `InvoiceRequest`.  So, a user MUST call `metadata_derived` if using `SigningPubkey::Derived` and otherwise MAY call `metadata_derived`, it seems.

> That metadata should probably be some kind of HMAC over the rest of the offer data plus a nonce.

Do you mean `metadata = HMAC(rest of offer) + nonce`? In which case, when responding, the nonce is used to derive the signing key after verifying the HMAC using the `ExpandedKey` and the appropriate offer TLV records, I assume? Let me know if I'm missing something.

jkczyz commented:
I was able to make the above sketched out change without using an `Option` for `OfferContents::signing_pubkey` by changing the `Derived` variant to `Derived(&'a ExpandedKey, Nonce)`. This let's us derive the signing_pubkey immediately and add the nonce later to the metadata. This also means `OfferBuilder::metadata_derived` doesn't need to be called when using this variant.

Looking for feedback on the `signing_pubkey` derivation, though, and earlier questions related to it. The latest push has it implemented.

jkczyz commented:
> I guess I'm confused - do we want the `OfferBuilder` to be tied to the LDK-specific offer metadata, or do we want to do that at a higher level? I had recalled previously that you said you wanted them to be separate.

Turns out it is much cleaner to incorporate it into the builder. But it still allows users to set it explicitly by giving `SigningPubkey::Explicit` and calling `OfferBuilder::metadata`. We could make dedicated methods and use `pub(crate)` where applicable, of course, if we want to hide it from the public interface.

jkczyz commented on lightning/src/ln/inbound_payment.rs:
```diff
@@ -63,6 +65,49 @@ impl ExpandedKey {
+	pub const LENGTH: usize = 16;
+
+	///
+	pub fn from_entropy_source<ES: EntropySource>(entropy_source: &ES) -> Self {
```
I thought about that but it gets kinda ugly as you'd need to pass it into `OfferBuilder::build` or have a type parameter on `SigningPubkey` even when not needed.

jkczyz commented on lightning/src/offers/offer.rs:
```diff
@@ -102,13 +120,21 @@ impl OfferBuilder {
 	///
 	/// Use a different pubkey per offer to avoid correlating offers.
-	pub fn new(description: String, signing_pubkey: PublicKey) -> Self {
+	pub fn new(description: String, signing_pubkey: SigningPubkey) -> Self {
```
When using the `Explicit` variant, the metadata will only be generated if `metadata_derived` is called. Two use cases:
- LDK-independent usage (must call `metadata`)
- LDK usage with public node id (must call `metadata_derived`)

When using `Derived`, the metadata will be generated without needing to call anything. See tests.

jkczyz replied:
Note that, as written, if a user gives `SigningPubkey::Explicit` and calls `OfferBuilder::metadata_derived` with an LDK-independent `ExpandedKey`, then `InvoiceRequest::verify` should work correctly. So it could be useful leaving this exposed for general usage.

jkczyz replied:
You'd have to ensure that the nonce used to create the signing pubkey is the same one passed to `OfferBuilder::metadata_derived`. With the current approach the builder ensures that it is the same.

jkczyz replied:
We'd have to make `OfferContents::metadata` an `Option` if we did so.

jkczyz replied:
> What if we instead move the `signing_pubkey` argument and the metadata type argument to the final `build` method? Then we could make it all one argument - either explicit for both or implicit for both.

Note that `metadata` is optional, so it isn't required when given an explicit `signing_pubkey`. Also, that approach doesn't support verification for when an explicit `signing_pubkey` is given (node id or otherwise), FWIW.

> Can we use a `BuildingOffer` struct or so?

Meh, not sure if I really want to repeat the fields and need to do something similar in all the other types, which may have nested contents. Seems error-prone.

> The current API just seems really confusing - users have to provide an enum variant at `OfferBuilder` constructor which can be explicit, in which case they can call the manual metadata function, or it can be derived, in which case they cant (or dont need to) call either metadata function, and if they provide derived they _cant_ call the derived-metadata function (unless its with the same data) or they'll get the wrong pubkey at the end?

Hmmm... if we think it will be confusing, then I'm more inclined to keep the constructor as it was and add a `pub(crate)` constructor for the derived variant taking `&ExpandedKey` and `Nonce`.

Then the question would be whether or not we want to support verification externally. If no, then `OfferBuilder::metadata_derived` is only needed for when used internally with a node id. So we can either make it `pub(crate)` or replace it with another `pub(crate)` constructor.

My feeling is that LDK-independent users likely would just want to set the `signing_pubkey` and `metadata` directly and won't care about the derived versions or verification. If true, I'd lean towards the `pub(crate)` approached outlined above.

jkczyz commented:
FYI, pushed some commits with clean-up and also using `SigningPubkey` for payer metadata in `InvoiceRequest`.

jkczyz commented:
While implementing message handling base on this PR, I came across a few decisions that need to be made. @TheBlueMatt @valentinewallace Let me know if you have any thoughts on the following.

### `ExpandedKey` ownership

`ExpandedKey` is needed at both creation and handling time for constructing and verifying the metadata, respectively. `ExpandedKey` is constructed from `KeyMaterial` retrieved from a `NodeSigner`. Currently, `ChannelManager` constructs and owns the `ExpandedKey`.

For builder creation, we can add methods to `ChannelManager` so that it retains `ExpandedKey` ownership. Since `ChannelManager` also has an `EntropySource`, it can create a `Nonce` and construct a `DerivedPubkey` for the builder.

```rust
pub fn create_offer_builder(&self, description: String) -> OfferBuilder {
    let nonce = inbound_payment::Nonce::from_entropy_source(&*self.entropy_source);
    let signing_pubkey = DerivedPubkey::new(&self.inbound_payment_key, nonce);
    OfferBuilder::deriving_signing_pubkey(description, signing_pubkey)
 }
```

But for verification, either `OnionMessenger` or its internal handler will need the `ExpandedKey`. A few options come to mind:
- reconstructed the `ExpandedKey` such that `OnionMessenger` or its internal handler has a copy
- have a shared reference, moving the creation of `ExpandedKey` out of `ChannelManager` entirely
- give `OnionMessenger` a reference to the `ChannelManager` for verification
- delegate verification to the user via an event instead, requiring them to call `ChannelManager` for verification

### Message handling flow

Onion messages containing either an `InvoiceRequest` or an `Invoice` need to be handled by sending an `Invoice` or making a payment, respectively, or by sending back an `InvoiceError`.

For `InvoiceRequest`, it can theoretically be verified and have an `Invoice` sent as a response automatically given we don't support creating offers with `Amount::Currency`.

For `Invoice`, the payment could be similarly sent without any user interaction. However, if the invoice is for an offer using `Amount::Currency` without an explicit amount set in the `InvoiceRequest`, then we may want to surface an `InvoiceReceived` event to the user to allow them to confirm the amount before making payment.

Note, also that handling will require constructing blinded reply paths or blinded payment paths, which indicates the need for a `ChannelManager` with a `Router`. Also, access to the `ExpandedKey` when constructing an `Invoice` and a manner to sign the `Invoice` using a `DerivedPubkey`. With that in mind, we may need to provide indication to the user of failure to create an `Invoice` as well (e.g., if a blinded path could not be formed). Could possibly use an event if automating?

The question here is then somewhat related to the `ExpandedKey` ownership, since it is needed when constructing an `Invoice`, too. Should we have the flow be event-centric, with user calling methods on `ChannelManager` / `OnionMessenger` when handling? `OnionMessenger` would then need to be an `EventsProvider`, IIUC.

valentinewallace commented:
> reconstructed the ExpandedKey such that OnionMessenger or its internal handler has a copy

This seems fine to me since `OnionMessenger` has a `node_signer` already.

> However, if the invoice is for an offer using Amount::Currency without an explicit amount set in the InvoiceRequest, then we may want to surface an InvoiceReceived event to the user to allow them to confirm the amount before making payment.

This sounds like a stopgap before real support for `Amount::Currency`, at which point we could remove it? It seems preferable to just refuse to pay such offers for the time being 🤔 

> With that in mind, we may need to provide indication to the user of failure to create an Invoice as well (e.g., if a blinded path could not be formed). Could possibly use an event if automating?

Maybe we could expand `PaymentFailed` to cover this?

jkczyz commented:
> Somehow I was thinking `ChannelManager` _would be_ that "internal handler" - that `OnionMessenger` would have a `Bolt12MessageHandler` trait, which `ChannelManager` would implement. Because invoices have to find their way to the `ChannelManager` for payment anyway this makes sense IMO - if we have an offer we want to pay, tell `ChannelManager`, it will generate and send the `invoice_request` as an `OnionMessageHandler` and then when it gets the `invoice` it already knows about it and can pay it.

Yeah, this seems like the right approach. Then the user can choose to implement their own handling if they aren't using the stateless verification. I have this implemented in #2039 for reference. The handling method will need to either take an `&OnionMessenger` so it can call `send_onion_message` in response or return a response for `OnionMessenger` to send for them, IIUC.

Still an open question is if we want to expose an `InvoiceReceived` event to the users for the `Currency::Amount` case, or fail to pay for now as @valentinewallace suggested.

jkczyz commented:
@valentinewallace @TheBlueMatt Have some questions / comments mostly related to the follow-up #2039, which now has the most recent push.

### Message signing

I added a `sign_bolt12_message` method to `NodeSigner` and additional parameters to the `sign` methods in the `offers` module, including the `metadata`. This is need to extract the `Nonce` in order to sign an `Invoice` for a particular `Offer`, for instance. `NodeSigner` has the `get_inbound_payment_key_material` method used to create the `ExpandedKey`, so `KeysManager` will need to recreate the `ExpandedKey` when signing (or store it on creation). That means we still need the `ExpandedKey` in two places.

### Blinded paths

I have utility methods for creating offers, etc. in `ChannelManager`, which will need to set blinded paths. Additionally, the `OffersMessageHandler` implementation for `ChannelManager` also will need to do so when creating an `Invoice` when handling an `InvoiceRequest` and will need `BlindedPayInfo` for each path.

Since blinded paths need to be at least two hops, we'll need a `NetworkGraph` at very least to construct these. This likely means `ChannelManager` needs another parameter. 😕 Or should we add methods to the `Router` trait? This isn't for finding a full path, only the blinded portion.

valentinewallace commented:
> Since blinded paths need to be at least two hops, we'll need a NetworkGraph at very least to construct these. This likely means ChannelManager needs another parameter. 😕 Or should we add methods to the Router trait? This isn't for finding a full path, only the blinded portion.

Conceptually, it would make sense to me to add method(s) to the `Router` trait.

jkczyz commented:
Rebased

jkczyz commented:
> If we're including a unique signing pk, cant we just use that as the metadata? ie take all the fields in the offer, hash them with the hmac, then treat that as the signing secret key and check that the signing pk in the offer matches. If it does, we have authenticated without needing the extra bytes in the offer/qr code.

IIUC, then we'd still need the nonce in the metadata but could drop the `Sha256Hash`. Or am I misunderstanding?

Also, FWIW, if we can't form a blinded path, we'd need to use the node pubkey instead of unique signing key. So in that case we couldn't do the check. That is, when using `OfferBuilder::new` and `OfferBuilder::metadata_derived` instead of only `OfferBuilder::deriving_signing_pubkey`.

jkczyz commented:
> If we're including a unique signing pk, cant we just use that as the metadata? ie take all the fields in the offer, hash them with the hmac, then treat that as the signing secret key and check that the signing pk in the offer matches. If it does, we have authenticated without needing the extra bytes in the offer/qr code.

Ugh, so this puts us back in the situation where we can't initialized `OfferContents::signing_pubkey` until all the fields are set, but we don't want it to be an `Option`.

jkczyz commented:
> Discussed offline, but, yea, we should consider a wrapper or a new builder internal.

I have this working for offer metadata and invoice request verification. See the two fixup commits on this branch: https://github.com/jkczyz/rust-lightning/commits/2023-01-stateless-offers-alt

Will need to do something similar when constructing `InvoiceRequest` and `Refund`. The verification logic on the second fixup is straightforward. But I'm hoping to get feedback on how the metadata and signing pubkey construction is formulated. I feel it could be a bit cleaner such that less responsibility is on the builder.

jkczyz commented:
> At a high level that looks fine to me. If we call with `deriving_signing_pubkey` why don't we just do `deriving_metadata_signing_pubkey` and skip the `metadata_derived` call entirely, calculating it in build?

That's what is happening now. It's an error to call `metadata_derived` after calling `deriving_signing_pubkey`. There are essentially three cases:
- `OfferBuilder::new` optionally followed by `OfferBuilder::metadata`
- `OfferBuilder::new` optionally followed by `OfferBuilder::metadata_derived` (internal only)
- `OfferBuilder::deriving_signing_pubkey` (internal only)

Maybe the naming could be refined. The last case is only including the nonce in the metadata and using the hmac for deriving the signing pubkey.

jkczyz commented:
Sorry about the delay on this. I ended up making a `OfferContentsWithoutSigningPubkey` struct to avoid repeating fields in `OfferBuilder`, which also made a bit cleaner. Will likely pull that and the similar change to `InvoiceRequestContents` out into separate commits, as I did with `RefundContents`.

jkczyz commented:
> Sorry about the delay on this. I ended up making a `OfferContentsWithoutSigningPubkey` struct to avoid repeating fields in `OfferBuilder`, which also made a bit cleaner. Will likely pull that and the similar change to `InvoiceRequestContents` out into separate commits, as I did with `RefundContents`.

Latest push has these clean-ups and squashes the fix-ups. Should be good for another pass.

jkczyz commented:
> > OfferBuilder::new optionally followed by OfferBuilder::metadata_derived (internal only)
> 
> Right, but why bother with calling `metadata_derived`, why don't we _just_ call `OfferBuilder::deriving_signing_pubkey` and then have `build` decide whether to derive the signing pubkey or not based on if there are blinded paths (and always derive the metadata). That leaves the driver with less thinking to do.

We'd need to pass an optional pubkey to build, so that it could be set if there are no blinded paths. This doesn't seem right given (1) it's for internal use and shouldn't affect the external interface and (2) it seems weird to call `deriving_signing_pubkey` but not actually derive a `signing_pubkey` in some cases. I'd rather have the interface do what it says rather than do something potentially surprising depending on the builder's state.

Additionally, as of now I have the utility return an `OfferBuilder`, so the user would need to call `build` with the right pubkey rather than letting `ChannelManager` do so. I guess we could pass the pubkey to `deriving_signing_pubkey`, too, instead? But that feels weird.

```rust
pub fn create_offer_builder(&self, description: String) -> OfferBuilder {
    let nonce = inbound_payment::Nonce::from_entropy_source(&*self.entropy_source);
    match self.create_blinded_paths(1) {
        Ok(paths) if !paths.is_empty() => {
            OfferBuilder::deriving_signing_pubkey(description, &self.inbound_payment_key, nonce)
                .path(paths.into_iter().next().unwrap())
            },
        Ok(_) | Err(_) => {
            let signing_pubkey = self.get_our_node_id();
            OfferBuilder::new(description, signing_pubkey)
                .metadata_derived(&self.inbound_payment_key, nonce).unwrap()
        },
    }
}
```

jkczyz commented on lightning/src/ln/inbound_payment.rs:
```diff
@@ -63,6 +65,49 @@ impl ExpandedKey {
+#[derive(Clone, Copy)]
+pub struct Nonce(pub(crate) [u8; Self::LENGTH]);
+
+impl Nonce {
```
It's kinda nice having a type even if it wraps a `u128` (though that adds a conversion when checking the metadata) as it encapsulates creation and properties like length and byte order. Otherwise, you'd have places in the code that need to get those right.

jkczyz commented on lightning/src/offers/invoice.rs:
```diff
@@ -469,8 +470,14 @@ impl Invoice {
 
+	/// Verifies that the invoice was for a request or refund created using the given key.
+	#[allow(unused)]
+	pub(crate) fn verify(&self, key: &ExpandedKey) -> bool {
```
In the upcoming PR, I have `NodeSigner::sign_bolt12_message` given the metadata, so it could extract it from there. Although, soon it will need to be decrypted first unless the signer could do that.

Otherwise, if returned here, we'd have callers pipe it into `UnsignedInvoiceRequest::sign`. Though, with the recent change for deriving the keys, I was considering leveraging the `Option<KeyPair>` field that I added to `UnsignedInvoiceRequest` instead of exposing the `Nonce`.

jkczyz commented:
> I guess I don't really see why either of these is important enough to have more logic on the builder's driver that is required to drive the builder correctly, rather than just making the builder's API "foolproof" and not requiring any driver specific behavior.

Ok, so this does clean it up quite a bit. And I'd be able to drop the commits making sub-structs for `OfferContents` and `RefundContents`. Though note that when creating an `InvoiceRequest` there are no blinded paths in the message itself. Instead, they are specified in the onion message's reply path, which is required to send back an invoice, IIUC. So I suppose we'd just have a third constructor for this case?

jkczyz commented:
> Ok, so this does clean it up quite a bit. And I'd be able to drop the commits making sub-structs for `OfferContents` and `RefundContents`. Though note that when creating an `InvoiceRequest` there are no blinded paths in the message itself. Instead, they are specified in the onion message's reply path, which is required to send back an invoice, IIUC. So I suppose we'd just have a third constructor for this case?

Added this third `InvoiceRequestBuilder` constructor. Also addressed TODO to encrypt the nonce.

valentinewallace commented on lightning/src/offers/offer.rs:
```diff
@@ -381,6 +436,44 @@ impl Offer {
 		Ok(InvoiceRequestBuilder::new(self, metadata, payer_id))
```
`request_invoice` docs should be updated to: s/Creates an [`InvoiceRequest`]/Creates a builder for an [`InvoiceRequest`], to match `Offer::new` and `Refund::new`

valentinewallace commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,137 @@
+/// Message metadata which possibly is derived from [`MetadataMaterial`] such that it can be
+/// verified.
+pub(super) enum Metadata {
+	Empty,
+	Bytes(Vec<u8>),
+	Derived(MetadataMaterial),
+	DerivedSigningPubkey(MetadataMaterial),
```
Could we document each variant?

valentinewallace commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,137 @@
+	}
+
+	if metadata.len() == EncryptedNonce::LENGTH {
+		let secp_ctx = Secp256k1::new();
```
Should this be passed in?

valentinewallace commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,137 @@
+	}
+
+	let nonce = EncryptedNonce::try_from(&metadata[..EncryptedNonce::LENGTH])
+		.unwrap()
```
Could return false instead of unwrapping

jkczyz reviewed (commented):
Comments addressed.

jkczyz commented:
FYI, I messed up the the last push. Should have a new push soon once I resolve some merge conflicts.

jkczyz commented:
> FYI, I messed up the the last push. Should have a new push soon once I resolve some merge conflicts.

All set now.

valentinewallace commented on lightning/src/offers/offer.rs:
```diff
@@ -204,13 +238,36 @@ impl OfferBuilder {
+		let (metadata, signing_pubkey) = self.metadata.into_parts();
+
+		if let Some(metadata) = metadata {
+			self.offer.metadata = Some(metadata);
```
I can't find the resolution to [this discussion](https://github.com/lightningdevkit/rust-lightning/pull/1989#issuecomment-1457185176). Why do we need to also set the metadata if we're deriving the signing pubkey?

jkczyz commented on lightning/src/offers/offer.rs:
```diff
@@ -222,10 +279,7 @@ impl OfferBuilder {
 	}
 
 	pub(super) fn build_unchecked(self) -> Offer {
```
Hmm... what names would you suggest? Note we probably shouldn't combine these since the code shouldn't be visible outside the file except for tests.

jkczyz commented on lightning/src/offers/offer.rs:
```diff
@@ -103,12 +106,38 @@ impl OfferBuilder {
+		}
+	}
+
+	/// Similar to [`OfferBuilder::new`] except, if [`OfferBuilder::path`] is not called, the
+	/// signing pubkey is derived from the given [`ExpandedKey`] and nonce. This provides recipient
```
Whoops, good catch. Fixed here and in `RefundBuilder`.

valentinewallace commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -81,18 +83,54 @@ const SIGNATURE_TAG: &'static str = concat!("lightning", "invoice_request", "sig
-			invoice_request: InvoiceRequestContents {
+			invoice_request: InvoiceRequestContentsWithoutPayerId {
 				payer: PayerContents(metadata), offer: offer.contents.clone(), chain: None,
 				amount_msats: None, features: InvoiceRequestFeatures::empty(), quantity: None,
-				payer_id, payer_note: None,
+				payer_note: None,
+			},
+			metadata: Metadata::Empty,
```
It's weird that we have two metadata fields in `InvoiceRequestBuilder` that aren't super explicitly disambiguated. Not sure if it'd be possible to pare down to one

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -82,21 +83,54 @@ const SIGNATURE_TAG: &'static str = concat!("lightning", "invoice_request", "sig
+	}
+
+	#[allow(unused)]
+	pub(super) fn deriving_metadata(
```
Added a helper to construct empty `InvoiceRequestContentsWithoutPayerId` from an `Offer`.

jkczyz commented on lightning/src/offers/refund.rs:
```diff
@@ -117,13 +121,43 @@ impl RefundBuilder {
+	/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+	/// [`ExpandedKey`]: crate::ln::inbound_payment::ExpandedKey
+	#[allow(unused)]
+	pub(crate) fn deriving_payer_id(
```
Missed this comment for the latest push. Let me think on a name. I suppose the important thing we want to convey is the resulting message can be verified in some way when given back later.

jkczyz replied:
Gonna punt on this until we can think of something that makes sense across all three builders.

valentinewallace commented on lightning/src/offers/offer.rs:
```diff
@@ -94,6 +97,7 @@ use std::time::SystemTime;
 /// [module-level documentation]: self
 pub struct OfferBuilder {
 	offer: OfferContents,
+	metadata: Metadata,
```
It would be nice if we could either unify or disambiguate this `metadata` field with `OfferContents::metadata`, similar for `InvoiceRequest`. At this point, it seems like a more accurate name for this would be `metadata_metadata`?

jkczyz replied:
> It would be nice if we could either unify or disambiguate this `metadata` field with `OfferContents::metadata`, similar for `InvoiceRequest`.

I took a stab at this in 7c48b4bb531290d779a5477b26c2fe0c73438a4e and 699550b771058de44bc065da02ac1fda17aaaa51 fixups. The primary issue in my mind is it makes it such that a constructed `Offer` must only use the `Metadata::Bytes` variant. So there are some assumptions that this is the case and an `unwrap` when verifying which otherwise would not be needed. Let me know what you and @TheBlueMatt think before I try the same for `InvoiceRequest` and `Refund`.

> At this point, it seems like a more accurate name for this would be `metadata_metadata`?

Not sure I follow this naming. It's more like metadata construction strategy or the like.

jkczyz replied:
Actually, the `unwrap` isn't necessary (29868983ec836a7828c18b780826dabf71ca455d) but the assumptions remain.

jkczyz replied:
> > which requires an `unwrap` in `InvoiceRequestContentsWithoutPayerId::metadata`: [d931c89](https://github.com/lightningdevkit/rust-lightning/commit/d931c890cc1527ee6d9690c8e5b788f1d20ebdd3).
> 
> Seems like we can do a (maybe cleaned up) version of this:
> 
> ```
> diff --git a/lightning/src/offers/invoice_request.rs b/lightning/src/offers/invoice_request.rs
> index b64c53ad3..a845bbab3 100644
> --- a/lightning/src/offers/invoice_request.rs
> +++ b/lightning/src/offers/invoice_request.rs
> @@ -485,6 +485,7 @@ impl InvoiceRequestContents {
> 
>  impl InvoiceRequestContentsWithoutPayerId {
>         pub fn metadata(&self) -> &[u8] {
> +               if self.payer.0.as_bytes().is_none() { return &[] }
>                 &self.payer.0.as_bytes().unwrap()[..]
>         }
> ```
> 

Heh, actually had done the same thing for the offer metadata in the follow-up PR in a different context lol.

> That said, I don't have a super strong opinion based on those diffs. It's nice to unify the field but if you think the code is nicer without, that sg. If we keep the PR as-is then I'd rather rename `metadata` to `metadata_strategy` or similar.

Ended up going with the unified approach.

valentinewallace commented on lightning/src/offers/offer.rs:
```diff
@@ -774,21 +887,106 @@ mod tests {
+			.request_invoice(vec![1; 32], payer_pubkey()).unwrap()
+			.build().unwrap()
+			.sign(payer_sign).unwrap();
+		assert!(!invoice_request.verify(&expanded_key, &secp_ctx));
```
We could check that the payer id/metadata are not factored into verification

jkczyz replied:
Hmmm... not possible. They are used when verifying, so altering them would cause a failure. I added some additional tests, though, testing for such failures.

valentinewallace reviewed (commented):
This basically lgtm, good time for @TheBlueMatt to come in

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -81,18 +83,48 @@ const SIGNATURE_TAG: &'static str = concat!("lightning", "invoice_request", "sig
+	}
+
+	#[allow(unused)]
+	pub(super) fn deriving_metadata(
```
> What is the use-case for paying with a known key?

Come to think of it, I'm not sure there is one. This was intended for the "can't find a blinded path" case for sending the onion message containing the `InvoiceRequest` with a reply path. But if we can't form a reply path, then we can't really send the request other than using our node id as the introduction node, I suppose.

> Does it make sense instead to be able to split the nonce into payer-id-nonce and per-payment-nonce and then derive both separately?

Maybe? I can't recall how the recurring payment case is suppose to work, honestly. Is there just one request from an offer and invoice sent each period? Or does the payer need to send another request each period?

jkczyz replied:
Renamed the parameter to `payer_id` and updated the docs.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,172 @@
+	}
+}
+
+impl io::Write for MetadataMaterial {
```
Gonna noodle on this a bit. Currently, `into_parts` returns `(Vec<u8>, Option<KeyPair>)`, but then we immediately go and make a `Metadata::Bytes` with the `Vec<u8>`. We probably should just convert it internally to a `Metadata::Bytes` instead of consuming now that `Metadata` is used throughout. Then we'd need to return the `KeyPair` if applicable.

jkczyz replied:
Cleaned this up quite a bit based on feedback. Now there is only a `derive_from` method taking a `Writeable` TLV stream.

jkczyz commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -170,8 +202,50 @@ impl<'a> InvoiceRequestBuilder<'a> {
+				inner: self.invoice_request,
+				payer_id,
+			},
+			keys,
```
If VLS is implementing `NodeSigner`, then it should have the `KeyMaterial` used to produce the `ExpandedKey`. Thus, given the `Invoice` bytes, it should be able to verify the metadata and produce the `KeyPair` needed for signing, right?

jkczyz replied:
Yeah, so note that the derived variation is now `pub`, FWIW. We could do something like:

```rust
pub fn build_and_sign(mut self) -> Result<Invoice, ...>
```

But now we need a special error type that can be both `SemanticError` and `SignError<E>`. So I'm a little more inclined to make a separate `sign` function unless you prefer a new error type.

Also, we should probably do something similar for `InvoiceBuilder`, but there we'll need to extract the keys from the metadata. We also need to determine *if* we should extract them or if the node id was used and sign accordingly with `NodeSigner`. I suppose that isn't too bad but we'll always need to pass a `sign` closure even if not used.

jkczyz replied:
Done. Also did the same for `OfferBuilder::metadata`. The bindings wrappers will need to hold a builder with a `dyn` parameter and we'll need to make the builder parameter `?Sized`, I believe.

jkczyz replied:
Oh, forgot about the `InvoiceBuilder`. I think we need another `InvoiceRequest::respond_with` method in order to have the compiler enforce correct signing by using a different type parameterization on the returned builder. So we'd need a dynamic check in that method on whether the `Offer` used a derived signing pubkey.

jkczyz replied:
Added in last commit.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,172 @@
+		bytes
+	}
+
+	fn into_parts(self) -> (Vec<u8>, KeyPair) {
```
Renamed as part of the related rework of `Metadata`.

jkczyz reviewed (commented):
Addressed some comments. Looking into the rest.

jkczyz commented:
Squashed fixups.

valentinewallace commented:
Looks minor but CI is sad

jkczyz commented:
> [ `f53420a`](/lightningdevkit/rust-lightning/commit/f53420a89172028b2818b79b2d71cfe52c23b180)

Fixed. `git diff-tree -U3  f53420a..a7741ebf`:

```diff
diff --git a/lightning/src/offers/invoice.rs b/lightning/src/offers/invoice.rs
index cceeb85c..f8b46613 100644
--- a/lightning/src/offers/invoice.rs
+++ b/lightning/src/offers/invoice.rs
@@ -1180,14 +1180,14 @@ mod tests {
 			.sign(payer_sign).unwrap();
 
 		if let Err(e) = invoice_request
-			.respond_with_using_derived_keys(payment_paths(), payment_hash()).unwrap()
+			.respond_with_using_derived_keys_no_std(payment_paths(), payment_hash(), now()).unwrap()
 			.build_and_sign(&expanded_key, &secp_ctx)
 		{
 			panic!("error building invoice: {:?}", e);
 		}
 
 		match invoice_request
-			.respond_with_using_derived_keys(payment_paths(), payment_hash()).unwrap()
+			.respond_with_using_derived_keys_no_std(payment_paths(), payment_hash(), now()).unwrap()
 			.build_and_sign(&ExpandedKey::new(&KeyMaterial([41; 32])), &secp_ctx)
 		{
 			Ok(_) => panic!("expected error"),
@@ -1203,7 +1203,7 @@ mod tests {
 			.sign(payer_sign).unwrap();
 
 		match invoice_request
-			.respond_with_using_derived_keys(payment_paths(), payment_hash()).unwrap()
+			.respond_with_using_derived_keys_no_std(payment_paths(), payment_hash(), now()).unwrap()
 			.build_and_sign(&expanded_key, &secp_ctx)
 		{
 			Ok(_) => panic!("expected error"),
diff --git a/lightning/src/offers/signer.rs b/lightning/src/offers/signer.rs
index 58e52d70..fd9d67c2 100644
--- a/lightning/src/offers/signer.rs
+++ b/lightning/src/offers/signer.rs
@@ -139,7 +139,7 @@ impl MetadataMaterial {
 
 /// Verifies data given in a TLV stream was used to produce the given metadata, consisting of:
 /// - a 128-bit [`EncryptedNonce`] and possibly
-/// - a [`Sha256Hash`] of the unencrypted nonce and the TLV records using the [`ExpandedKey`].
+/// - a [`Sha256`] hash of the unencrypted nonce and the TLV records using the [`ExpandedKey`].
 ///
 /// If the latter is not included in the metadata, the TLV stream is used to check if the given
 /// `signing_pubkey` can be derived from it.
```

jkczyz commented:
Oops, missed some. `git diff-tree -U3  a7741ebf..c0877d6e`:

```diff
diff --git a/lightning/src/offers/invoice.rs b/lightning/src/offers/invoice.rs
index f8b46613..0b162cb7 100644
--- a/lightning/src/offers/invoice.rs
+++ b/lightning/src/offers/invoice.rs
@@ -557,8 +557,7 @@ impl Invoice {
 	}
 
 	/// Verifies that the invoice was for a request or refund created using the given key.
-	#[allow(unused)]
-	pub(crate) fn verify<T: secp256k1::Signing>(
+	pub fn verify<T: secp256k1::Signing>(
 		&self, key: &ExpandedKey, secp_ctx: &Secp256k1<T>
 	) -> bool {
 		self.contents.verify(TlvStream::new(&self.bytes), key, secp_ctx)
diff --git a/lightning/src/offers/invoice_request.rs b/lightning/src/offers/invoice_request.rs
index 37d71028..d52b9d85 100644
--- a/lightning/src/offers/invoice_request.rs
+++ b/lightning/src/offers/invoice_request.rs
@@ -496,7 +496,7 @@ impl InvoiceRequest {
 	/// Creates an [`InvoiceBuilder`] for the request using the given required fields and that uses
 	/// derived signing keys from the originating [`Offer`] to sign the [`Invoice`].
 	///
-	/// See [`InvoiceBuilder::respond_with`] for further details.
+	/// See [`InvoiceRequest::respond_with`] for further details.
 	///
 	/// [`Invoice`]: crate::offers::invoice::Invoice
 	#[cfg(feature = "std")]
@@ -513,7 +513,7 @@ impl InvoiceRequest {
 	/// Creates an [`InvoiceBuilder`] for the request using the given required fields and that uses
 	/// derived signing keys from the originating [`Offer`] to sign the [`Invoice`].
 	///
-	/// See [`InvoiceBuilder::respond_with_no_std`] for further details.
+	/// See [`InvoiceRequest::respond_with_no_std`] for further details.
 	///
 	/// [`Invoice`]: crate::offers::invoice::Invoice
 	pub fn respond_with_using_derived_keys_no_std(
@@ -528,8 +528,7 @@ impl InvoiceRequest {
 	}
 
 	/// Verifies that the request was for an offer created using the given key.
-	#[allow(unused)]
-	pub(crate) fn verify<T: secp256k1::Signing>(
+	pub fn verify<T: secp256k1::Signing>(
 		&self, key: &ExpandedKey, secp_ctx: &Secp256k1<T>
 	) -> bool {
 		self.contents.inner.offer.verify(&self.bytes, key, secp_ctx)
diff --git a/lightning/src/offers/offer.rs b/lightning/src/offers/offer.rs
index d25f7a74..7ff51251 100644
--- a/lightning/src/offers/offer.rs
+++ b/lightning/src/offers/offer.rs
@@ -148,10 +148,11 @@ impl OfferBuilder<DerivedMetadata> {
 	/// by using a different signing pubkey for each offer, assuming a different nonce is used.
 	/// Otherwise, the provided `node_id` is used for the signing pubkey.
 	///
-	/// Also, sets the metadata when [`OfferBuilder::build`] is called such that it can be used to
-	/// verify that an [`InvoiceRequest`] was produced for the offer given an [`ExpandedKey`].
+	/// Also, sets the metadata when [`OfferBuilder::build`] is called such that it can be used by
+	/// [`InvoiceRequest::verify`] to determine if the request was produced for the offer given an
+	/// [`ExpandedKey`].
 	///
-	/// [`InvoiceRequest`]: crate::offers::invoice_request::InvoiceRequest
+	/// [`InvoiceRequest::verify`]: crate::offers::invoice_request::InvoiceRequest::verify
 	/// [`ExpandedKey`]: crate::ln::inbound_payment::ExpandedKey
 	pub fn deriving_signing_pubkey<ES: Deref>(
 		description: String, node_id: PublicKey, expanded_key: &ExpandedKey, entropy_source: ES
@@ -460,6 +461,8 @@ impl Offer {
 	///
 	/// Useful to protect the sender's privacy.
 	///
+	/// [`InvoiceRequest::payer_id`]: crate::offers::invoice_request::InvoiceRequest::payer_id
+	/// [`InvoiceRequest::metadata`]: crate::offers::invoice_request::InvoiceRequest::metadata
 	/// [`Invoice::verify`]: crate::offers::invoice::Invoice::verify
 	/// [`ExpandedKey`]: crate::ln::inbound_payment::ExpandedKey
 	pub fn request_invoice_deriving_payer_id<ES: Deref>(
@@ -479,6 +482,8 @@ impl Offer {
 	/// [`InvoiceRequest::payer_id`] instead deriving a different key for each request.
 	///
 	/// Useful for recurring payments using the same `payer_id` with different invoices.
+	///
+	/// [`InvoiceRequest::payer_id`]: crate::offers::invoice_request::InvoiceRequest::payer_id
 	pub fn request_invoice_deriving_metadata<ES: Deref>(
 		&self, payer_id: PublicKey, expanded_key: &ExpandedKey, entropy_source: ES
 	) -> Result<InvoiceRequestBuilder<ExplicitPayerId>, SemanticError>
```

valentinewallace commented on lightning/src/offers/offer.rs:
```diff
@@ -87,30 +92,86 @@ use crate::prelude::*;
+	///
+	/// [`InvoiceRequest::verify`]: crate::offers::invoice_request::InvoiceRequest::verify
+	/// [`ExpandedKey`]: crate::ln::inbound_payment::ExpandedKey
+	pub fn deriving_signing_pubkey<ES: Deref>(
```
It seems like a nicer api to have a separate method that doesn't require a `node_id` (that'll be unused in the majority of cases, iiuc). Could you remind me exactly why we can't do that? Is it to avoid an `Option`/`OfferWithoutNodeId` struct internally? Too many GH comments, sorry..

jkczyz commented on lightning/src/ln/inbound_payment.rs:
```diff
@@ -48,21 +48,114 @@ pub struct ExpandedKey {
+		hmac
+	}
+
+	pub(crate) fn encrypt_nonce_for_offer(
```
Technically, it would still be exposed as it would be in the metadata. Would you still rather we not encrypt it?

jkczyz replied:
Note that his will also effectively remove the message-specific IV bytes. Should those be used elsewhere instead?

jkczyz replied:
Done in latest push.

valentinewallace commented on lightning/src/offers/refund.rs:
```diff
@@ -317,27 +372,26 @@ impl Refund {
 		&self, payment_paths: Vec<(BlindedPath, BlindedPayInfo)>, payment_hash: PaymentHash,
 		signing_pubkey: PublicKey,
-	) -> Result<InvoiceBuilder, SemanticError> {
+	) -> Result<InvoiceBuilder<ExplicitSigningPubkey>, SemanticError> {
```
Has there been discussion about how we'll verify an incoming payment as originating from our own invoice? I assume it'll be via the `path_id`. Looking at the spec, I think `signing_pubkey` isn't used here and can be random (the spec says that payers should verify it "out-of-band" in the refund case).

jkczyz replied:
Yeah, I believe it is by using the path id for a payment secret.

Ok, I had to slightly rework the part of the builder for extracting keys from offer metadata to sign the invoice. But the newly added `Refund::respond_using_derived_keys` and it are now fairly seamless on the `InvoiceBuilder` side.

valentinewallace commented:
![image](https://user-images.githubusercontent.com/10647624/230968340-d066ed9e-1a62-422b-a145-d962afef2681.png)

FYI I made a flowchart

valentinewallace reviewed (commented):
Nothing major, I think this is almost good to land? Fine with squash/rebase.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	fn derive_metadata_and_keys(self) -> (Vec<u8>, KeyPair) {
+		let secp_ctx = Secp256k1::new();
```
Would you care if a context is cloned for this? Ideally, it is passed when the builder is created and carried along either in the applicable `Metadata` variants or directly in `MetadataMaterial`. Otherwise, if a reference is kept, the lifetime would trickle up to the top-level messages (i.e., `Offer`, `Refund`, etc.).

An alternative would be to pass it in to the `build` calls. But then either some will take a context that is never used or we further make separate `build` methods for the derived variants. I kinda wanted to avoid passing to `build` though because that means the `ChannelManager` utilities for creating builders in the next PR won't return fully usable builders; the caller would need to fetch the context when eventually calling `build`.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,205 @@
+			} else {
+				false
+			},
+			Metadata::Derived(_) => false,
```
Eh, this is a consequence of using `Metadata` in `Offer`, etc. It implements `PartialEq `, and so should its parts. But it never makes sense to have a derived variant once built. I think it's only needed in test assertions, so we could conditionally define it, I suppose, both here and in all the messages.

jkczyz replied:
Eh, that takes us full circle where the builder needs to hang on to the `Metadata`. Not sure if I want to re-do all that work as it would substantially affect the derivation code in the builders. Made it test-only instead.

jkczyz reviewed (commented):
Haven't pushed the updates but had some open questions in the comments.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,205 @@
+		let derived_pubkey = SecretKey::from_slice(hmac.as_inner()).unwrap().public_key(&secp_ctx);
+		fixed_time_eq(&signing_pubkey.serialize(), &derived_pubkey.serialize())
+	} else {
+		fixed_time_eq(&metadata[Nonce::LENGTH..], &hmac.into_inner())
```
Fixed.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,130 @@
+			Metadata::Bytes(_) => (self, None),
+			Metadata::Derived(mut metadata_material) => {
+				tlv_stream.write(&mut metadata_material.hmac).unwrap();
+				(Metadata::Bytes(metadata_material.derive_metadata()), None)
```
I'd rather keep this in `MetadataMaterial` for symmetry with the other case.

jkczyz reviewed (commented):
I believe that I addressed all comments, but let me know if I missed anything.

valentinewallace commented on lightning/src/offers/invoice_request.rs:
```diff
@@ -162,56 +214,119 @@ impl<'a> InvoiceRequestBuilder<'a> {
-		Ok(UnsignedInvoiceRequest { offer, invoice_request })
+		let unsigned_invoice_request = self.build_without_checks();
+
+		if unsigned_invoice_request.invoice_request.inner.payer.0.as_bytes().is_none() {
```
The spec requires it to be set to some random bytes, in the case of `invoice_requests`. Seems like we should check, don't care if we do it in this PR though

jkczyz commented:
Rebased and squashed

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,232 @@
+
+use crate::prelude::*;
+
+const DERIVED_METADATA_HMAC_INPUT: &[u8; 15] = b"derive_metadata";
```
Fixed

jkczyz commented on lightning/src/offers/offer.rs:
```diff
@@ -204,28 +261,48 @@ impl OfferBuilder {
+				let mut tlv_stream = self.offer.as_tlv_stream();
+				debug_assert_eq!(tlv_stream.metadata, None);
+				tlv_stream.metadata = None;
+				tlv_stream.node_id = None;
```
Fixed.

jkczyz commented on lightning/src/offers/offer.rs:
```diff
@@ -483,6 +613,27 @@ impl OfferContents {
+				let tlv_stream = TlvStream::new(bytes).range(OFFER_TYPES).filter(|record| {
+					match record.r#type {
+						OFFER_METADATA_TYPE => false,
+						OFFER_NODE_ID_TYPE => !self.metadata.as_ref().unwrap().derives_keys(),
```
`Metadata::derives_keys` was updated in the first commit that introduced verification to check the length of the metadata.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,240 @@
+
+/// Extracts the derived keys from `metadata` that was created from the given TLV stream and
+/// [`ExpandedKey`].
+pub(super) fn extract_derived_keys<'a, T: secp256k1::Signing>(
```
Removed

valentinewallace reviewed (dismissed):
LGTM, feel free to address nits in follow-up

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -0,0 +1,226 @@
+
+	pub fn derives_keys(&self) -> bool {
+		match self {
+			Metadata::Bytes(bytes) => bytes.len() == Nonce::LENGTH,
```
Added a comment. This is simply to determine which fields should be included when verifying a message. User-set bytes even if length `Nonce::LENGTH` would still fail verification, as would be expected.
